File: packages/server/src/workspaces/runtime.ts
import { ChildProcess, spawn, spawnSync } from "child_process"
import { existsSync, statSync } from "fs"
import path from "path"
import { EventBus } from "../events/bus"
import { LogLevel, WorkspaceLogEntry } from "../api-types"
import { Logger } from "../logger"
import { buildSpawnSpec, buildWslSignalSpec } from "./spawn"

const SENSITIVE_ENV_KEY = /(PASSWORD|TOKEN|SECRET)/i
const WSL_PID_MARKER = "__EMBEDDEDCOWORK_WSL_PID__:"

function redactEnvironment(env: Record<string, string | undefined>): Record<string, string | undefined> {
  const redacted: Record<string, string | undefined> = {}
  for (const [key, value] of Object.entries(env)) {
    if (value === undefined) {
      redacted[key] = value
      continue
    }
    redacted[key] = SENSITIVE_ENV_KEY.test(key) ? "[REDACTED]" : value
  }
  return redacted
}

interface LaunchOptions {
  workspaceId: string
  folder: string
  binaryPath: string
  environment?: Record<string, string>
  logLevel?: string
  onExit?: (info: ProcessExitInfo) => void
}

export interface ProcessExitInfo {
  workspaceId: string
  code: number | null
  signal: NodeJS.Signals | null
  requested: boolean
}

interface ManagedProcess {
  child: ChildProcess
  requestedStop: boolean
  wsl?: {
    distro: string
    linuxPid: number | null
  }
}

export class WorkspaceRuntime {
  private processes = new Map<string, ManagedProcess>()

  constructor(private readonly eventBus: EventBus, private readonly logger: Logger) {}

  async launch(options: LaunchOptions): Promise<{ pid: number; port: number; exitPromise: Promise<ProcessExitInfo>; getLastOutput: () => string }> {
    this.validateFolder(options.folder)

    const logLevel = typeof options.logLevel === "string" ? options.logLevel.toUpperCase() : "DEBUG"
    const args = ["serve", "--port", "0", "--print-logs", "--log-level", logLevel]
    const env = { ...process.env, ...(options.environment ?? {}) }

    let exitResolve: ((info: ProcessExitInfo) => void) | null = null
    const exitPromise = new Promise<ProcessExitInfo>((resolveExit) => {
      exitResolve = resolveExit
    })

    // Store recent output for debugging - keep last 50 lines from each stream
    const MAX_OUTPUT_LINES = 50
    const recentStdout: string[] = []
    const recentStderr: string[] = []
    const getLastOutput = () => {
      const combined: string[] = []
      if (recentStderr.length > 0) {
        combined.push("Error Stream")
        combined.push(...recentStderr.slice(-10))
      }
      if (recentStdout.length > 0) {
        combined.push("Output Stream")
        combined.push(...recentStdout.slice(-10))
      }
      return combined.join("\n")
    }

    return new Promise((resolve, reject) => {
      const propagatedEnvKeys = Object.keys(options.environment ?? {})
      const spec = buildSpawnSpec(options.binaryPath, args, {
        cwd: options.folder,
        env,
        propagateEnvKeys: propagatedEnvKeys,
        wslPidMarker: WSL_PID_MARKER,
      })
      const commandLine = [spec.command, ...spec.args].join(" ")
      this.logger.info(
        {
          workspaceId: options.workspaceId,
          folder: options.folder,
          binary: options.binaryPath,
          spawnCommand: spec.command,
          commandLine,
        },
        "Launching OpenCode process",
      )

      this.logger.debug(
        {
          workspaceId: options.workspaceId,
          spawnArgs: spec.args,
        },
        "OpenCode spawn args",
      )

      this.logger.trace(
        {
          workspaceId: options.workspaceId,
          env: redactEnvironment(env),
        },
        "OpenCode spawn environment",
      )
      const detached = process.platform !== "win32"
      const child = spawn(spec.command, spec.args, {
        cwd: spec.cwd,
        env: spec.env,
        stdio: ["ignore", "pipe", "pipe"],
        detached,
        ...spec.options,
      })

      const managed: ManagedProcess = {
        child,
        requestedStop: false,
        ...(spec.wsl ? { wsl: { distro: spec.wsl.distro, linuxPid: null } } : {}),
      }
      this.processes.set(options.workspaceId, managed)

      let stdoutBuffer = ""
      let stderrBuffer = ""
      let portFound = false

      let warningTimer: NodeJS.Timeout | null = null

      const startWarningTimer = () => {
        warningTimer = setInterval(() => {
          this.logger.warn({ workspaceId: options.workspaceId }, "Workspace runtime has not reported a port yet")
        }, 10000)
      }

      const stopWarningTimer = () => {
        if (warningTimer) {
          clearInterval(warningTimer)
          warningTimer = null
        }
      }

      startWarningTimer()

      const cleanupStreams = () => {
        stopWarningTimer()
        child.stdout?.removeAllListeners()
        child.stderr?.removeAllListeners()
      }

      const handleExit = (code: number | null, signal: NodeJS.Signals | null) => {
        this.logger.info({ workspaceId: options.workspaceId, code, signal }, "OpenCode process exited")
        this.processes.delete(options.workspaceId)
        cleanupStreams()
        child.removeListener("error", handleError)
        child.removeListener("exit", handleExit)
        const exitInfo: ProcessExitInfo = {
          workspaceId: options.workspaceId,
          code,
          signal,
          requested: managed.requestedStop,
        }
        if (exitResolve) {
          exitResolve(exitInfo)
          exitResolve = null
        }
        if (!portFound) {
          const recentOutput = getLastOutput().trim()
          const reason = recentOutput || stderrBuffer || `Process exited with code ${code}`
          reject(new Error(reason))
        } else {
          options.onExit?.(exitInfo)
        }
      }

      const handleError = (error: Error) => {
        cleanupStreams()
        child.removeListener("exit", handleExit)
        this.processes.delete(options.workspaceId)
        this.logger.error({ workspaceId: options.workspaceId, err: error }, "Workspace runtime error")
        if (exitResolve) {
          exitResolve({ workspaceId: options.workspaceId, code: null, signal: null, requested: managed.requestedStop })
          exitResolve = null
        }
        reject(error)
      }

      child.on("error", handleError)
      child.on("exit", handleExit)

      child.stdout?.on("data", (data: Buffer) => {
        const text = data.toString()
        stdoutBuffer += text
        const lines = stdoutBuffer.split("\n")
        stdoutBuffer = lines.pop() ?? ""

        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed) continue

          if (managed.wsl && trimmed.startsWith(WSL_PID_MARKER)) {
            const linuxPid = Number.parseInt(trimmed.slice(WSL_PID_MARKER.length), 10)
            if (Number.isFinite(linuxPid) && linuxPid > 0) {
              managed.wsl.linuxPid = linuxPid
              this.logger.debug({ workspaceId: options.workspaceId, linuxPid }, "Captured WSL OpenCode PID")
            }
            continue
          }

          recentStdout.push(trimmed)
          if (recentStdout.length > MAX_OUTPUT_LINES) {
            recentStdout.shift()
          }

          this.emitLog(options.workspaceId, "info", line)

          if (!portFound) {
            const portMatch = line.match(/opencode server listening on http:\/\/.+:(\d+)/i)
            if (portMatch) {
              portFound = true
              stopWarningTimer()
              child.removeListener("error", handleError)
              const port = parseInt(portMatch[1], 10)
              this.logger.info({ workspaceId: options.workspaceId, port }, "Workspace runtime allocated port")
              resolve({ pid: child.pid!, port, exitPromise, getLastOutput })
            }
          }
        }
      })

      child.stderr?.on("data", (data: Buffer) => {
        const text = data.toString()
        stderrBuffer += text
        const lines = stderrBuffer.split("\n")
        stderrBuffer = lines.pop() ?? ""

        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed) continue

          recentStderr.push(trimmed)
          if (recentStderr.length > MAX_OUTPUT_LINES) {
            recentStderr.shift()
          }

          this.emitLog(options.workspaceId, "error", line)
        }
      })
    })
  }

  async stop(workspaceId: string): Promise<void> {
    const managed = this.processes.get(workspaceId)
    if (!managed) return

    managed.requestedStop = true
    const child = managed.child
    this.logger.info({ workspaceId }, "Stopping OpenCode process")

    const pid = child.pid
    if (!pid) {
      this.logger.warn({ workspaceId }, "Workspace process missing PID; cannot stop")
      return
    }

    const isAlreadyExited = () => child.exitCode !== null || child.signalCode !== null

    const tryKillPosixGroup = (signal: NodeJS.Signals) => {
      try {
        // Negative PID targets the process group (POSIX).
        process.kill(-pid, signal)
        return true
      } catch (error) {
        const err = error as NodeJS.ErrnoException
        if (err?.code === "ESRCH") {
          return true
        }
        this.logger.debug({ workspaceId, pid, err }, "Failed to signal POSIX process group")
        return false
      }
    }

    const tryKillSinglePid = (signal: NodeJS.Signals) => {
      try {
        process.kill(pid, signal)
        return true
      } catch (error) {
        const err = error as NodeJS.ErrnoException
        if (err?.code === "ESRCH") {
          return true
        }
        this.logger.debug({ workspaceId, pid, err }, "Failed to signal workspace PID")
        return false
      }
    }

    const tryTaskkill = (force: boolean) => {
      const args = ["/PID", String(pid), "/T"]
      if (force) {
        args.push("/F")
      }

      try {
        const result = spawnSync("taskkill", args, { encoding: "utf8" })
        const exitCode = result.status
        if (exitCode === 0) {
          return true
        }
        // If the PID is already gone, treat it as success.
        const stderr = (result.stderr ?? "").toString().toLowerCase()
        const stdout = (result.stdout ?? "").toString().toLowerCase()
        const combined = `${stdout}\n${stderr}`
        if (combined.includes("not found") || combined.includes("no running instance") || combined.includes("process") && combined.includes("not")) {
          return true
        }
        this.logger.debug({ workspaceId, pid, exitCode, stderr: result.stderr, stdout: result.stdout }, "taskkill failed")
        return false
      } catch (error) {
        this.logger.debug({ workspaceId, pid, err: error }, "taskkill failed to execute")
        return false
      }
    }

    const trySignalWslProcess = (signal: NodeJS.Signals) => {
      if (process.platform !== "win32" || !managed.wsl?.linuxPid) {
        return false
      }

      try {
        const spec = buildWslSignalSpec(managed.wsl.distro, managed.wsl.linuxPid, signal)
        const result = spawnSync(spec.command, spec.args, { encoding: "utf8" })
        const exitCode = result.status
        if (exitCode === 0) {
          return true
        }

        const stderr = (result.stderr ?? "").toString().toLowerCase()
        const stdout = (result.stdout ?? "").toString().toLowerCase()
        const combined = `${stdout}\n${stderr}`
        if (combined.includes("no such process") || combined.includes("not found")) {
          return true
        }

        this.logger.debug(
          { workspaceId, pid, linuxPid: managed.wsl.linuxPid, distro: managed.wsl.distro, exitCode, stderr: result.stderr, stdout: result.stdout },
          "WSL kill failed",
        )
        return false
      } catch (error) {
        this.logger.debug({ workspaceId, pid, linuxPid: managed.wsl.linuxPid, distro: managed.wsl.distro, err: error }, "WSL kill failed to execute")
        return false
      }
    }

    const sendStopSignal = (signal: NodeJS.Signals) => {
      if (process.platform === "win32") {
        // WSL-backed launches need a Linux signal first because the tracked Windows PID belongs to wsl.exe.
        if (!trySignalWslProcess(signal)) {
          // Fallback to the Windows process tree rooted at pid. Use /F only for escalation.
          tryTaskkill(signal === "SIGKILL")
        }
        return
      }

      // Prefer process-group signaling so wrapper launchers (bun/node) don't orphan the real server.
      const groupOk = tryKillPosixGroup(signal)
      if (!groupOk) {
        // Fallback to direct PID kill.
        tryKillSinglePid(signal)
      }
    }

    await new Promise<void>((resolve, reject) => {
      let escalationTimer: NodeJS.Timeout | null = null

      const cleanup = () => {
        child.removeListener("exit", onExit)
        child.removeListener("error", onError)
        if (escalationTimer) {
          clearTimeout(escalationTimer)
          escalationTimer = null
        }
      }

      const onExit = () => {
        cleanup()
        resolve()
      }
      const onError = (error: Error) => {
        cleanup()
        reject(error)
      }

      if (isAlreadyExited()) {
        this.logger.debug({ workspaceId, exitCode: child.exitCode, signal: child.signalCode }, "Process already exited")
        cleanup()
        resolve()
        return
      }

      child.once("exit", onExit)
      child.once("error", onError)

      this.logger.debug(
        { workspaceId, pid, detached: process.platform !== "win32" },
        "Sending SIGTERM to workspace process (tree/group)",
      )
      sendStopSignal("SIGTERM")

      escalationTimer = setTimeout(() => {
        escalationTimer = null
        if (isAlreadyExited()) {
          this.logger.debug({ workspaceId, pid }, "Workspace exited before SIGKILL escalation")
          return
        }
        this.logger.warn({ workspaceId, pid }, "Process did not stop after SIGTERM, escalating")
        sendStopSignal("SIGKILL")
      }, 2000)
    })
  }

  private emitLog(workspaceId: string, level: LogLevel, message: string) {
    const entry: WorkspaceLogEntry = {
      workspaceId,
      timestamp: new Date().toISOString(),
      level,
      message: message.trim(),
    }

    this.eventBus.publish({ type: "workspace.log", entry })
  }

  private validateFolder(folder: string) {
    const resolved = path.resolve(folder)
    if (!existsSync(resolved)) {
      throw new Error(`Folder does not exist: ${resolved}`)
    }
    const stats = statSync(resolved)
    if (!stats.isDirectory()) {
      throw new Error(`Path is not a directory: ${resolved}`)
    }
  }
}

File: packages/server-rust/src/workspaces/runtime.rs
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;
use tokio::io::AsyncBufReadExt;
use tokio::process::{Child, Command};
use tokio::sync::{Mutex, oneshot};

use crate::api_types::{LogLevel, WorkspaceLogEntry};
use crate::events::bus::EventBus;
use crate::logger::Logger;
use crate::workspaces::spawn::build_spawn_spec;

const PORT_DETECTION_TIMEOUT_SECS: u64 = 30;

#[derive(Debug, Clone)]
pub struct ProcessExitInfo {
    pub workspace_id: String,
    pub code: Option<i32>,
    pub signal: Option<String>,
    pub requested: bool,
}

struct ManagedProcess {
    child: Arc<Mutex<Child>>,
    requested_stop: bool,
}

pub struct WorkspaceRuntime {
    processes: Arc<Mutex<HashMap<String, ManagedProcess>>>,
    event_bus: EventBus,
    logger: Logger,
}

impl WorkspaceRuntime {
    pub fn new(event_bus: EventBus, logger: Logger) -> Self {
        Self {
            processes: Arc::new(Mutex::new(HashMap::new())),
            event_bus,
            logger,
        }
    }

    pub async fn launch(
        &self,
        options: LaunchOptions,
    ) -> Result<LaunchResult, String> {
        self.validate_folder(&options.folder)?;

        let log_level = options
            .log_level
            .as_deref()
            .unwrap_or("DEBUG")
            .to_uppercase();

        let args = vec![
            "serve".to_string(),
            "--port".to_string(),
            "0".to_string(),
            "--print-logs".to_string(),
            "--log-level".to_string(),
            log_level,
        ];

        let empty_env = HashMap::new();
        let env = options.environment.as_ref().unwrap_or(&empty_env);

        let spec = build_spawn_spec(
            &options.binary_path,
            &args,
            &options.folder,
            env,
        );

        let command_line = format!("{} {}", spec.command, spec.args.join(" "));
        tracing::info!(
            component = %self.logger.component,
            workspace_id = %options.workspace_id,
            folder = %options.folder,
            binary = %options.binary_path,
            spawn_command = %spec.command,
            command_line = %command_line,
            "Launching OpenCode process"
        );

        let mut cmd = Command::new(&spec.command);
        cmd.args(&spec.args)
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped());

        if let Some(ref dir) = spec.current_dir {
            cmd.current_dir(dir);
        }
        if let Some(ref extra_env) = spec.env {
            cmd.envs(extra_env);
        }

        let mut child = cmd
            .spawn()
            .map_err(|e| format!("Failed to spawn OpenCode process: {}", e))?;

        let pid = child.id().unwrap_or(0);
        let workspace_id = options.workspace_id.clone();

        tracing::info!(
            component = %self.logger.component,
            workspace_id = %workspace_id,
            pid = pid,
            "OpenCode process spawned, waiting for port"
        );

        // Take stdout for port detection
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| "Failed to capture stdout from OpenCode process".to_string())?;

        // Spawn background task to read stderr (prevents buffer deadlock)
        if let Some(stderr) = child.stderr.take() {
            let ws_id = workspace_id.clone();
            let logger = self.logger.child("opencode-stderr");
            let event_bus = self.event_bus.clone();
            tokio::spawn(async move {
                let reader = tokio::io::BufReader::new(stderr);
                let mut lines = reader.lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    let trimmed = line.trim().to_string();
                    if trimmed.is_empty() {
                        continue;
                    }
                    tracing::debug!(
                        component = %logger.component,
                        workspace_id = %ws_id,
                        "[opencode:stderr] {}",
                        trimmed
                    );
                    event_bus.publish(WorkspaceEventPayload::WorkspaceLog {
                        event_type: "workspace.log".to_string(),
                        entry: WorkspaceLogEntry {
                            workspace_id: ws_id.clone(),
                            timestamp: chrono::Utc::now().to_rfc3339(),
                            level: LogLevel::Error,
                            message: trimmed,
                        },
                    });
                }
            });
        }

        // Port detection from stdout
        let (port_tx, port_rx) = oneshot::channel::<u16>();

        let ws_id_for_stdout = workspace_id.clone();
        let logger_for_stdout = self.logger.child("opencode-stdout");
        let event_bus_for_stdout = self.event_bus.clone();
        tokio::spawn(async move {
            let reader = tokio::io::BufReader::new(stdout);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let trimmed = line.trim().to_string();
                if trimmed.is_empty() {
                    continue;
                }

                tracing::trace!(
                    component = %logger_for_stdout.component,
                    workspace_id = %ws_id_for_stdout,
                    "[opencode:stdout] {}",
                    trimmed
                );

                event_bus_for_stdout.publish(WorkspaceEventPayload::WorkspaceLog {
                    event_type: "workspace.log".to_string(),
                    entry: WorkspaceLogEntry {
                        workspace_id: ws_id_for_stdout.clone(),
                        timestamp: chrono::Utc::now().to_rfc3339(),
                        level: LogLevel::Info,
                        message: trimmed.clone(),
                    },
                });

                // Detect port from: "opencode server listening on http://...:PORT"
                if let Some(pos) = trimmed.to_lowercase().find("opencode server listening on http://") {
                    let rest = &trimmed[pos + 34..]; // skip past "opencode server listening on http://"
                    if let Some(colon_pos) = rest.rfind(':') {
                        let port_str = rest[colon_pos + 1..]
                            .trim_end_matches(|c: char| !c.is_ascii_digit());
                        if let Ok(port) = port_str.parse::<u16>() {
                            tracing::info!(
                                component = %logger_for_stdout.component,
                                workspace_id = %ws_id_for_stdout,
                                port = %port,
                                "Detected OpenCode listening port"
                            );
                            let _ = port_tx.send(port);
                            return;
                        }
                    }
                }
            }
        });

        // Wait for port detection with timeout
        let port = tokio::time::timeout(
            std::time::Duration::from_secs(PORT_DETECTION_TIMEOUT_SECS),
            port_rx,
        )
        .await
        .map_err(|_| {
            format!(
                "OpenCode did not report a listening port within {} seconds",
                PORT_DETECTION_TIMEOUT_SECS
            )
        })?
        .map_err(|_| "Failed to read OpenCode port from stdout (channel closed)".to_string())?;

        let local_host = "127.0.0.1".to_string();

        let managed = ManagedProcess {
            child: Arc::new(Mutex::new(child)),
            requested_stop: false,
        };

        self.processes
            .lock()
            .await
            .insert(options.workspace_id.clone(), managed);

        tracing::info!(
            component = %self.logger.component,
            workspace_id = %workspace_id,
            pid = pid,
            port = port,
            "OpenCode process started successfully"
        );

        Ok(LaunchResult {
            pid,
            port,
            exit_promise: Box::pin(async move { None }),
            get_last_output: Arc::new(|| String::new()),
            local_host,
        })
    }

    pub async fn stop(&self, workspace_id: &str) -> Result<(), String> {
        let mut processes = self.processes.lock().await;
        if let Some(mut managed) = processes.remove(workspace_id) {
            managed.requested_stop = true;
            // Explicitly kill the child process (tokio::process::Child does
            // NOT auto-kill on drop).
            let mut child = managed.child.lock().await;
            let _ = child.start_kill();
        }
        Ok(())
    }

    pub async fn shutdown(&self) {
        let mut processes = self.processes.lock().await;
        for (_, managed) in processes.iter() {
            let mut child = managed.child.lock().await;
            let _ = child.start_kill();
        }
        processes.clear();
    }

    fn validate_folder(&self, folder: &str) -> Result<(), String> {
        let path = Path::new(folder);
        if !path.exists() {
            return Err(format!("Folder does not exist: {}", folder));
        }
        if !path.is_dir() {
            return Err(format!("Path is not a directory: {}", folder));
        }
        Ok(())
    }
}

use crate::api_types::WorkspaceEventPayload;

pub struct LaunchOptions {
    pub workspace_id: String,
    pub folder: String,
    pub binary_path: String,
    pub environment: Option<HashMap<String, String>>,
    pub log_level: Option<String>,
}

pub struct LaunchResult {
    pub pid: u32,
    pub port: u16,
    pub exit_promise: std::pin::Pin<
        Box<dyn std::future::Future<Output = Option<ProcessExitInfo>> + Send>,
    >,
    pub get_last_output: Arc<dyn Fn() -> String + Send + Sync>,
    pub local_host: String,
}
